File: scripts/update-pixiv-weekly-top10.mjs
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, "..");
const outputPath = path.join(projectRoot, "src", "data", "pixiv-weekly-top10.json");
const imageDir = path.join(projectRoot, "public", "pixiv-weekly");

const endpoint =
  "https://www.pixiv.net/ranking.php?mode=weekly&content=illust&format=json";

const cookie = process.env.PIXIV_COOKIE?.trim();

const headers = {
  Accept: "application/json, text/plain, */*",
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
  Referer: "https://www.pixiv.net/ranking.php?mode=weekly&content=illust",
};

if (cookie) {
  headers.Cookie = cookie;
}

const toProxyImageUrl = (url) => {
  if (!url || typeof url !== "string") return "";
  return `https://wsrv.nl/?url=${encodeURIComponent(url)}&w=720&output=webp`;
};

const toPixivReUrl = (url) => {
  if (!url || typeof url !== "string") return "";
  // i.pximg.net needs referer; i.pixiv.re is a public proxy mirror.
  return url
    .replace("https://i.pximg.net/", "https://i.pixiv.re/")
    .replace(/\/c\/\d+x\d+\//, "/");
};

const toPixivCatUrl = (id) => {
  if (!id) return "";
  return `https://pixiv.cat/${id}.jpg`;
};

const getExtFromUrl = (url) => {
  const clean = url.split("?")[0];
  const match = clean.match(/\.([a-zA-Z0-9]+)$/);
  return match ? match[1].toLowerCase() : "jpg";
};

const downloadToLocalImage = async (id, sources) => {
  await mkdir(imageDir, { recursive: true });

  for (const source of sources) {
    try {
      const res = await fetch(source, {
        headers: {
          Referer: "https://www.pixiv.net/",
          "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        },
      });
      if (!res.ok) continue;

      const contentType = String(res.headers.get("content-type") ?? "").toLowerCase();
      if (!contentType.startsWith("image/")) continue;

      const ext = getExtFromUrl(source);
      const filename = `${id}.${ext}`;
      const filepath = path.join(imageDir, filename);
      const buf = Buffer.from(await res.arrayBuffer());
      await writeFile(filepath, buf);
      return `/pixiv-weekly/${filename}`;
    } catch {
      // Try next source.
    }
  }

  return "";
};

const normalizeItem = async (item, index) => {
  const id = String(item.illust_id ?? item.id ?? "");
  const originalThumb = String(item.url ?? "");
  const tags = Array.isArray(item.tags)
    ? item.tags.map((tag) => String(tag)).filter(Boolean)
    : [];
  const fallbackThumbnailUrls = [toProxyImageUrl(originalThumb), toPixivReUrl(originalThumb), toPixivCatUrl(id)].filter(
    Boolean,
  );
  const localThumbnailUrl = await downloadToLocalImage(id, fallbackThumbnailUrls);

  return {
    rank: Number(item.rank ?? index + 1),
    id,
    title: String(item.title ?? ""),
    userName: String(item.user_name ?? ""),
    userId: String(item.user_id ?? ""),
    artworkUrl: id ? `https://www.pixiv.net/artworks/${id}` : "",
    thumbnailUrl: localThumbnailUrl || fallbackThumbnailUrls[0] || "",
    fallbackThumbnailUrls,
    originalThumbnailUrl: originalThumb,
    tags,
  };
};

const main = async () => {
  const response = await fetch(endpoint, { headers });
  if (!response.ok) {
    throw new Error(`Failed to fetch ranking: ${response.status} ${response.statusText}`);
  }

  const payload = await response.json();
  const contents = Array.isArray(payload?.contents) ? payload.contents : [];
  const top10 = (await Promise.all(contents.slice(0, 10).map(normalizeItem))).filter((item) => item.id);

  if (top10.length === 0) {
    throw new Error("No ranking items found. PIXIV_COOKIE may be required.");
  }

  const data = {
    generatedAt: new Date().toISOString(),
    source: endpoint,
    total: top10.length,
    items: top10,
  };

  await mkdir(path.dirname(outputPath), { recursive: true });
  await writeFile(outputPath, `${JSON.stringify(data, null, 2)}\n`, "utf8");
  console.log(`Updated ${outputPath} with ${top10.length} items.`);
};

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: scripts/update_pixiv_weekly_top10.py
#!/usr/bin/env python3
import json
import os
from pathlib import Path
from urllib.parse import urlparse
from datetime import datetime, timezone

from pixivpy3 import AppPixivAPI


ROOT = Path(__file__).resolve().parent.parent
OUTPUT_PATH = ROOT / "src" / "data" / "pixiv-weekly-top10.json"
IMAGE_DIR = ROOT / "public" / "pixiv-weekly"
SOURCE = "pixivpy3:illust_ranking(mode=week)"
TARGET_COUNT = 10
PAGE_SIZE = 30


def is_illustration(illust) -> bool:
    # Pixiv types are usually: "illustration", "manga", "ugoira"
    illust_type = str(getattr(illust, "type", "") or "").lower()
    return illust_type in {"illustration", "illust"}


def ext_from_url(url: str) -> str:
    path = urlparse(url).path
    suffix = Path(path).suffix.lower()
    if suffix in {".jpg", ".jpeg", ".png", ".webp"}:
        return suffix
    return ".jpg"


def main() -> None:
    refresh_token = (os.getenv("PIXIV_REFRESH_TOKEN") or "").strip()
    if not refresh_token:
        raise RuntimeError("PIXIV_REFRESH_TOKEN is missing.")

    IMAGE_DIR.mkdir(parents=True, exist_ok=True)
    OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)

    api = AppPixivAPI()
    api.set_accept_language("zh-cn")
    api.auth(refresh_token=refresh_token)

    collected = []
    seen_ids = set()
    offset = 0
    max_rounds = 8

    for _ in range(max_rounds):
        ranking = api.illust_ranking(mode="week", offset=offset)
        round_items = list(getattr(ranking, "illusts", []))
        if not round_items:
            break

        for illust in round_items:
            illust_id = str(getattr(illust, "id", "") or "")
            if not illust_id or illust_id in seen_ids:
                continue
            if not is_illustration(illust):
                continue
            seen_ids.add(illust_id)
            collected.append(illust)
            if len(collected) >= TARGET_COUNT:
                break

        if len(collected) >= TARGET_COUNT:
            break
        offset += PAGE_SIZE

    illusts = collected[:TARGET_COUNT]
    if not illusts:
        raise RuntimeError("No weekly ranking items returned from Pixiv API.")

    items = []
    for index, illust in enumerate(illusts, start=1):
        illust_id = str(illust.id)
        title = str(illust.title or "")
        user_name = str(getattr(illust.user, "name", "") or "")
        user_id = str(getattr(illust.user, "id", "") or "")

        image_urls = getattr(illust, "image_urls", None)
        thumb_url = ""
        if image_urls:
            thumb_url = (
                getattr(image_urls, "large", "")
                or getattr(image_urls, "medium", "")
                or getattr(image_urls, "square_medium", "")
                or ""
            )
        thumb_url = str(thumb_url)

        local_thumb = ""
        if thumb_url:
            ext = ext_from_url(thumb_url)
            filename = f"{illust_id}{ext}"
            try:
                api.download(thumb_url, path=str(IMAGE_DIR), name=filename)
                local_thumb = f"/pixiv-weekly/{filename}"
            except Exception:
                local_thumb = ""

        tags = []
        for tag in list(getattr(illust, "tags", [])):
            tag_name = str(getattr(tag, "name", "") or "").strip()
            if tag_name:
                tags.append(tag_name)

        items.append(
            {
                "rank": index,
                "id": illust_id,
                "title": title,
                "userName": user_name,
                "userId": user_id,
                "artworkUrl": f"https://www.pixiv.net/artworks/{illust_id}",
                "thumbnailUrl": local_thumb or thumb_url,
                "fallbackThumbnailUrls": [thumb_url] if local_thumb else [],
                "originalThumbnailUrl": thumb_url,
                "tags": tags,
            }
        )

    payload = {
        "generatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "source": SOURCE,
        "total": len(items),
        "items": items,
    }

    OUTPUT_PATH.write_text(
        json.dumps(payload, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    print(f"Updated {OUTPUT_PATH} with {len(items)} items.")


if __name__ == "__main__":
    main()
